### PR TITLE
Keymap for Web

### PR DESCRIPTION
**Pull request by mdebbar**
## Description

Implement correct key mapping for the web.

## Related Issues

This PR, along with https://github.com/flutter/flutter/pull/40186, enable tabbing between form fields on the web (https://github.com/flutter/flutter/issues/41396).

## Tests

I added tests in the following files:

* `test/services/raw_keyboard_test.dart`

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/services/keyboard_maps.dart [hunk 1]
```diff
@@ -1493,3 +1493,230 @@ const Map<int, PhysicalKeyboardKey> kLinuxToPhysicalKey = <int, PhysicalKeyboard
...
+  96: LogicalKeyboardKey.numpad0,
+  110: LogicalKeyboardKey.numpadDecimal,
+  93: LogicalKeyboardKey.contextMenu,
+  16: LogicalKeyboardKey.shift,
+  17: LogicalKeyboardKey.control,
+  18: LogicalKeyboardKey.alt,
+  91: LogicalKeyboardKey.meta,
+};
+
+// TODO(mdebbar): FINISH!
```
Comment (yjbanov) [inline packages/flutter/lib/src/services/keyboard_maps.dart]: Let's leave a link to an issue? https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-checking-in-comments-that-ask-questions

File: packages/flutter/lib/src/services/raw_keyboard_web.dart [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+import 'keyboard_key.dart';
+import 'keyboard_maps.dart';
+import 'raw_keyboard.dart';
+
+/// Platform-specific key event data for Web.
+///
+/// See also:
+///
+///  * [RawKeyboard], which uses this interface to expose key data.
+class RawKeyEventDataWeb extends RawKeyEventData {
```
Comment (yjbanov) [inline packages/flutter/lib/src/services/raw_keyboard_web.dart]: Can this be made `@immutable`?

File: packages/flutter/lib/src/services/raw_keyboard_web.dart [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  ///  * [isMetaPressed], to see if a META key is pressed.
+  final int metaState;
+
+  @override
+  final String keyLabel;
+
+  @override
+  PhysicalKeyboardKey get physicalKey {
+    if (kWebToPhysicalKey.containsKey(keyCode)) {
+      return kWebToPhysicalKey[keyCode] ?? PhysicalKeyboardKey.none;
```
Comment (yjbanov) [inline packages/flutter/lib/src/services/raw_keyboard_web.dart]: Is it possible to have a key with a `null` value? If not, then this one line is sufficient as a complete implementation of this getter.

File: packages/flutter/lib/src/services/raw_keyboard_web.dart [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+      case ModifierKey.numLockModifier:
+        return metaState & modifierNumLock != 0;
+      case ModifierKey.capsLockModifier:
+        return metaState & modifierCapsLock != 0;
+      case ModifierKey.scrollLockModifier:
+        return metaState & modifierScrollLock != 0;
+      case ModifierKey.functionModifier:
+      case ModifierKey.symbolModifier:
+        // On Web, the browser doesn't report the state of the FN and SYM modifiers.
+        return false;
```
Comment (yjbanov) [inline packages/flutter/lib/src/services/raw_keyboard_web.dart]: Can you have `default` together with fucntionModifier and symbolModifier and avoid two returns?

**Review by yjbanov — APPROVED** [review]

File: packages/flutter/lib/src/services/keyboard_maps.dart (hunk 1 above)
Comment (mdebbar) [reply packages/flutter/lib/src/services/keyboard_maps.dart]: Oops! This should be removed now.

File: packages/flutter/lib/src/services/keyboard_maps.dart [hunk 5]
```diff
@@ -1493,3 +1493,230 @@ const Map<int, PhysicalKeyboardKey> kLinuxToPhysicalKey = <int, PhysicalKeyboard
   0x00000080: PhysicalKeyboardKey.showAllWindows,
 };
 
+/// Maps Web-specific key codes to the matching [LogicalKeyboardKey].
```
Comment (gspencergoog) [inline packages/flutter/lib/src/services/keyboard_maps.dart]: Check out the comment at the top of the file: this is a generated file, generated from the Chrome source code.  See https://github.com/flutter/flutter/tree/master/dev/tools/gen_keycodes/README.md

Comment (mdebbar) [reply packages/flutter/lib/src/services/keyboard_maps.dart]: The comment was collapsed with the copyright comment so I didn't notice it :(

Thanks for pointing this out. I modified the code gen tool to produce the mappings for the web. Please take another look.

File: packages/flutter/lib/src/services/raw_keyboard_web.dart [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+    // Look to see if the [code] is one we know about and have a mapping for.
+    final LogicalKeyboardKey newKey = kWebToLogicalKey[code];
+    if (newKey != null) {
+      return newKey;
+    }
+
+    // This is a non-printable key that we don't know about, so we mint a new
+    // code with the autogenerated bit set.
+    const int webKeyIdPlane = 0x00900000000;
```
Comment (mdebbar) [inline packages/flutter/lib/src/services/raw_keyboard_web.dart]: @gspencergoog is there a method for assigning these numbers?

Comment (gspencergoog) [reply packages/flutter/lib/src/services/raw_keyboard_web.dart]: Yes, see https://github.com/flutter/flutter/blob/master/dev/tools/gen_keycodes/README.md

There's a section in the key id schema that has a table.  The code for web is `0x00800000000`.

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

It's good to see that using the Chromium headers pays off here: we can just use their names for things.

**Pull request merged**